### PR TITLE
feat(store): reusable states

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: 
https://github.com/ngxs/store/issues/179
https://github.com/ngxs/store/issues/741
https://github.com/ngxs/store/issues/719

## What is the new behavior?

throw an error, since the previous state will be overwritten

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/store/src/utils/store-validators.ts [hunk 1]
```diff
@@ -14,4 +17,33 @@ export abstract class StoreValidators {
...
+    const statesNames: Map<StateName, StateClassName> = new Map();
+
+    // https://jsperf.com/for-vs-foreach/75
+    for (let i = 0, size: number = states.length; i < size; i++) {
+      const state: StateClass = states[i]!;
+      const meta: MetaDataModel = this.validateStateMeta(state);
+      const stateName: string = <string>meta!.name;
+
+      if (statesNames.has(stateName)) {
+        const previousStateName: string = <string>statesNames.get(stateName);
```
Comment (arturovt) [inline packages/store/src/utils/store-validators.ts]:
```TS
const previousStateName = stateNames.get(stateName) as string;
```

File: packages/store/src/utils/store-validators.ts [hunk 2]
```diff
@@ -14,4 +17,33 @@ export abstract class StoreValidators {
...
   }
+
+  public static validateStateNames(states: StateClass[]): Map<StateName, StateClassName> {
+    const statesNames: Map<StateName, StateClassName> = new Map();
+
+    // https://jsperf.com/for-vs-foreach/75
+    for (let i = 0, size: number = states.length; i < size; i++) {
+      const state: StateClass = states[i]!;
+      const meta: MetaDataModel = this.validateStateMeta(state);
+      const stateName: string = <string>meta!.name;
```
Comment (arturovt) [inline packages/store/src/utils/store-validators.ts]:
```TS
const stateName = meta!.name as string;
```

File: packages/store/src/utils/store-validators.ts (hunk 1 above)
Comment (splincode) [reply packages/store/src/utils/store-validators.ts]: done

**Comment by splincode** [comment]
@markwhithfeld ping

File: tsconfig.json [hunk 3]
```diff
@@ -15,6 +15,7 @@
     "moduleResolution": "node",
     "emitDecoratorMetadata": true,
     "experimentalDecorators": true,
+    "downlevelIteration": true,
```
[the remark this answers is not in the document]
Comment (splincode) [reply tsconfig.json]: Issue: https://github.com/Microsoft/TypeScript/issues/27008

I use merge two maps
https://github.com/ngxs/store/pull/791/files#diff-62bb81999aabeb9aaa84c63a01b0670eR64

**Comment by splincode** [comment]
@markwhitfeld @arturovt I have improved the formation of the state tree so that the hmr-plugin will now work perfectly

File: tsconfig.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (arturovt) [reply tsconfig.json]: @markwhitfeld in simple terms `downlevelIteration` option is useful when you target `es5` and you have to do some iteration over ES2015 collection. TypeScript compiler emits optimized function that contains contains a proper implementation of the iteration protocol.

File: packages/store/src/internal/state-factory.ts [hunk 4]
```diff
@@ -49,83 +50,98 @@ export class StateFactory {
...
+  }
+
+  private static immutableDefaults(defaults: any): any {
+    let value = {};
+
+    if (Array.isArray(defaults)) {
+      value = [...defaults];
+    } else if (isObject(defaults)) {
+      value = { ...defaults };
+    } else if (defaults === undefined) {
```
Comment (arturovt) [inline packages/store/src/internal/state-factory.ts]: do you need this condition?

File: packages/store/src/internal/state-factory.ts [hunk 5]
```diff
@@ -195,4 +211,21 @@ export class StateFactory {
...
+  }
+
+  /**
+   * @description
+   * the method checks if the state has already been added to the tree
+   * and completed the life cycle
+   * @param name
+   * @param depth
+   */
+  private hasWillMounted(name: string, depth: string): boolean {
```
Comment (arturovt) [inline packages/store/src/internal/state-factory.ts]: `hasBeenMounted`

File: packages/store/src/utils/store-validators.ts [hunk 6]
```diff
@@ -14,4 +17,32 @@ export abstract class StoreValidators {
       throw new Error(this.stateNameErrorMessage(name));
     }
   }
+
+  public static getValidStateNamesMap(states: StateClass[]): Map<StateName, StateClassName> {
+    const uniqueStateNamesMap: Map<StateName, StateClassName> = new Map();
+
+    for (let i = 0, size: number = states.length; i < size; i++) {
```
Comment (arturovt) [inline packages/store/src/utils/store-validators.ts]: maybe call it `length` - more semantic maybe? and also do you need to specify the `number` type here?

File: packages/store/src/internal/state-factory.ts (hunk 4 above)
Comment (splincode) [reply packages/store/src/internal/state-factory.ts]: Why not?

Comment (arturovt) [reply packages/store/src/internal/state-factory.ts]: i mean you declare a `value` variable before all these conditions and assign to `{}` and also in this condition you assign to `{}`, maybe better:
```TS
let value = {};

if (Array.isArray(defaults)) {
  value = [...defaults];
} else if (isObject(defaults)) {
  value = { ...defaults };
} else if (defaults !== undefined) {
  value = defaults;
}
```

Comment (markwhitfeld) [reply packages/store/src/internal/state-factory.ts]: Just my 2c, I would say that having multiple returns would actually be better here. 
Guard clauses are one of the definite exceptions to the guidelines from Code Complete.
Could be expressed as:
```TS
if (Array.isArray(defaults)) return [...defaults];
if (isObject(defaults)) return { ...defaults };
if (typeof defaults === 'undefined') return {};
return defaults;
```

File: packages/store/src/utils/store-validators.ts [hunk 7]
```diff
@@ -14,4 +17,32 @@ export abstract class StoreValidators {
       throw new Error(this.stateNameErrorMessage(name));
     }
   }
+
+  public static getValidStateNamesMap(states: StateClass[]): Map<StateName, StateClassName> {
```
[the remark this answers is not in the document]
Comment (markwhitfeld) [reply packages/store/src/utils/store-validators.ts]: Would not detect duplicates from feature states.

File: tsconfig.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (markwhitfeld) [reply tsconfig.json]: I see that the reason this would have been added was to iterate over the `Map`.

**Pull request merged**